language/stdlib/modules/0L/ValidatorUniverse.move

```diff
@@ -31,21 +36,103 @@ address 0x1 {
       if (!in_set) {
         Vector::push_back<address>(&mut state.validators, addr);
+        unjail(sender);
+      }
+    }
+
+    // Permissions: Public, VM Only
+    public fun remove_validator_vm(vm: &signer, validator: address) acquires ValidatorUniverse {
+      assert(Signer::address_of(vm) == CoreAddresses::LIBRA_ROOT_ADDRESS(), 220101014010);
+
+      let state = borrow_global_mut<ValidatorUniverse>(CoreAddresses::LIBRA_ROOT_ADDRESS());
+      let (in_set, index) = Vector::index_of<address>(&state.validators, &validator);
+      if (in_set) {
+        Vector::remove<address>(&mut state.validators, index);
+      }
+    }
+
+    // Permissions: Public, Anyone.
+    // Can only remove self from validator list.
+    public fun remove_self(validator: &signer) acquires ValidatorUniverse {
+      let val = Signer::address_of(validator);
+      let state = borrow_global_mut<ValidatorUniverse>(CoreAddresses::LIBRA_ROOT_ADDRESS());
+      let (in_set, index) = Vector::index_of<address>(&state.validators, &val);
+      if (in_set) {
+        Vector::remove<address>(&mut state.validators, index);
       }
     }
 
     // A simple public function to query the EligibleValidators.
     // Function code: 03 Prefix: 220103
-    public fun get_eligible_validators(vm: &signer) : vector<address> acquires ValidatorUniverse {
+    public fun get_eligible_validators(vm: &signer): vector<address> acquires ValidatorUniverse {
       assert(Signer::address_of(vm) == CoreAddresses::LIBRA_ROOT_ADDRESS(), 220101014010);
       let state = borrow_global<ValidatorUniverse>(CoreAddresses::LIBRA_ROOT_ADDRESS());
       *&state.validators
     }
+
+    // Is a candidate for validation
+    public fun is_in_universe(miner: address): bool acquires ValidatorUniverse {
+      let state = borrow_global<ValidatorUniverse>(CoreAddresses::LIBRA_ROOT_ADDRESS());
+      Vector::contains<address>(&state.validators, &miner)
+    }
+
+    public fun jail(vm: &signer, validator: address) acquires JailedBit{
+      assert(Signer::address_of(vm) == CoreAddresses::LIBRA_ROOT_ADDRESS(), 220101014010);
+
+      borrow_global_mut<JailedBit>(validator).is_jailed = true;
+    }
+
+    public fun unjail_self(sender: &signer) acquires JailedBit {
+      // only a validator can un-jail themselves.
+      let validator = Signer::address_of(sender);
+      // check the node has been mining before unjailing.
+      assert(MinerState::node_above_thresh(sender, validator), 220102014010);
+      unjail(sender);
+    }
+
+    fun unjail(sender: &signer) acquires JailedBit {
+      let addr = Signer::address_of(sender);
+      if (!exists<JailedBit>(addr)) {
+        move_to<JailedBit>(sender, JailedBit{
+          is_jailed: false
+        });
+      };
+
+      borrow_global_mut<JailedBit>(addr).is_jailed = false;
+    }
+
+    public fun is_jailed(validator: address): bool acquires JailedBit {
+      if (!exists<JailedBit>(validator)) {
+        return false
```

Comment:
Well this is a query that can happen at any moment. Effectively it will migrate any account that is in an undefined state. For safety this should be false I think.

Reply:
Yeah, you're right. Perhaps in the future, once a quorum of validators have set the bit it could be switched, but it must be default off now.